src/base/utils/version.h

```diff
@@ -0,0 +1,184 @@
+        {
+            static_assert(N >= 3, "Array size is too small");
+            return (*this)[3];
+        }
+
+        ComponentType operator[](std::size_t i) const
+        {
+            return m_components.at(i);
+        }
+
+        operator QString() const
+        {
+            // find the last one non-zero component
+            std::size_t lastSignificantIndex = N - 1;
+            while (lastSignificantIndex >= 0 && (*this)[lastSignificantIndex] == 0)
+                --lastSignificantIndex;
+
+            if (lastSignificantIndex + 1 < Mandatory)     // lastSignificantIndex >= 0
+                lastSignificantIndex = Mandatory - 1;     // and Mandatory >= 1
+
+            QString res = QString::number((*this)[0]);
+            for (std::size_t i = 1; i <= lastSignificantIndex; ++i)
+                res += QLatin1Char('.') + QString::number((*this)[i]);
+            return res;
+        }
+
+        bool operator==(const ThisType &other) const
+        {
+            return m_components == other.m_components;
+        }
+
+        bool operator<(const ThisType &other) const
+        {
+            return m_components < other.m_components;
+        }
+
+        bool operator>(const ThisType &other) const
+        {
+            return m_components > other.m_components;
+        }
+
+        template <class AQString>
+        static Version tryParse(const AQString &s, const Version &defaultVersion)
+        {
+            try {
+                return Version(s);
+            }
+            catch (std::runtime_error &er) {
+                qDebug() << "Error parsing version:" << er.what();
+                return defaultVersion;
+            }
+        }
+
+    private:
+
+        using ComponentsArray = std::array<T, N>;
+
+        template <class QtStringsList>
+        static ComponentsArray parseList(const QtStringsList &versionParts)
+        {
+            if ((static_cast<std::size_t>(versionParts.size()) > N)
+                || (static_cast<std::size_t>(versionParts.size()) < Mandatory))
+                throw std::runtime_error ("Incorrect number of version components");
+
+            bool ok = false;
+            ComponentsArray res;
+            for (std::size_t i = 0; i < static_cast<std::size_t>(versionParts.size()); ++i) {
+                res[i] = static_cast<T>(versionParts[i].toInt(&ok));
+                if (!ok)
+                    throw std::runtime_error("Can not parse version component");
+            }
+            return res;
+        }
+
+        template <class QtStringsList>
+        Version(const QtStringsList &versionParts)
+            : m_components (parseList(versionParts)) // GCC 4.8.4 has problems with the uniform initializer here
+        {
+        }
+
+        ComponentsArray m_components;
+    };
+
+    template <class T, std::size_t N, std::size_t Mandatory>
+    inline bool operator!=(const Version<T, N, Mandatory> &left, const Version<T, N, Mandatory> &right)
```

Comment:
>With this one it is needed to read the function body,

and so is reading `==` and other relational operators, no?

> while operator outside of the class suggests that it does not have access to the non-public members.

Yes, but why this hint is important? It would make sense if implicit conversion are involved as I read here: http://stackoverflow.com/a/9554578, or doing https://github.com/qbittorrent/qBittorrent/pull/5264#discussion_r65396829.
And `!=` is a `const` method, it clearly hints it won't modify anything.

--
>@Chocobo1, do you suggest to make it a class member? Why, if so?

Can't really say I have a special consideration. IMO it's just simplicity.
If I see a `==` in a class declaration, then naturally (for me) there should be a `!=` nearby, if it's declared outside of the class then it must be doing something special, but now, seems not....
and I don't think we need to have `==` operator for different template instances now...

I'm lazy person (well, just a bit) so I'm found of [You Arent Gonna Need It](http://c2.com/xp/YouArentGonnaNeedIt.html) principle.
Nice discussing with you though.